### PR TITLE
message body parts should be a sub body

### PR DESCRIPTION
**Pull request by willpower232**
Resolves https://github.com/atech/postal/issues/375

### CODE REVIEW

File: app/models/outgoing_message_prototype.rb [hunk 1]
```diff
@@ -160,17 +160,15 @@ def raw_message
       mail.sender = @sender
       mail.subject = @subject
       mail.reply_to = @reply_to
-      if @html_body.blank? && attachments.empty?
-        mail.body = @plain_body
```
Comment (adamcrown) [inline app/models/outgoing_message_prototype.rb]: It seems like this outer condition that checks for plain text only messages should still exist. There's no reason to use `multipart/alternative` for those, right?

Comment (willpower232) [reply app/models/outgoing_message_prototype.rb]: I was initially erring on the side of caution but also the comment I linked to from the issue suggests that if you include a text/plain attachment without this arrangement, the attachment can get lost on some platforms.

Either way, its better to clearly define the parts within a message as theres no knowing what email software and servers will be involved in the passage of the message to its recipient.

**Comment by alexander-rieder** [comment]
when will this get merged?

**Comment by alexander-rieder** [comment]
Any news?

File: app/models/outgoing_message_prototype.rb (hunk 1 above)
Comment (sgregori) [reply app/models/outgoing_message_prototype.rb]: Works like a charm !

**Comment by jaydrogers** [comment]
I also discovered this issue and am interested in seeing this get merged.

**Comment by alexander-rieder** [comment]
@naft-a  What is stopping this pull request from getting merged?

**Comment by adamcooke** [comment]
I'd say the main thing holding this up is that we likely don't want plain text only e-mails to be within the multipart/alternative block.

* If there's a plain text only, it should just have that.
* If there's HTML only it should just have that.
* If there's both it should use multipart/alternative.

**Comment by catphish** [comment]
I've just tested this, and it doesn't appear that this patch is necessary. The current behaviour using the current mail gem appears to already be what this pull request aims to create:

```
irb(main):002:0> mail = Mail.new

irb(main):003:0> mail.text_part = Mail::Part.new
irb(main):004:0> mail.text_part.body = "arf"

irb(main):005:0> mail.html_part = Mail::Part.new
irb(main):006:0> mail.html_part.content_type = "text/html; charset=UTF-8"
irb(main):007:0> mail.html_part.body = 'html here'

irb(main):008:0> puts mail.to_s
```

```
Date: Thu, 05 Dec 2019 17:04:35 +0000
Message-ID: <5de938a34dfe_1d646b90a471586@postal.mail>
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_5de938625b425_1d646b90a4714ab";
 charset=UTF-8
Content-Transfer-Encoding: 7bit


----==_mimepart_5de938625b425_1d646b90a4714ab
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

arf
----==_mimepart_5de938625b425_1d646b90a4714ab
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

html here
----==_mimepart_5de938625b425_1d646b90a4714ab--
```

**Comment by catphish** [comment]
@alexander-rieder Are you seeing some behaviour that implies that this is not working correctly?
@willpower232 I won't merge this as-is because it breaks the ability to send a normal plain-text-only email. If necessary, we can split it out to create a multipart message when both parts are specified, but as far as I can see, the mail gem (at least now) does this automatically.

**Comment by Josh-G** [comment]
> The current behaviour using the current mail gem appears to already be what this pull request aims to create

@catphish not quite - your example doesn't involve attachments which is the crux of the issue.
I've created a little example to help demonstrate the issue at hand. Please bear with me as I'm not familiar with Ruby at all.

I've created a [gist with a minimal repro](https://gist.github.com/Josh-G/110c3617494b4e5d10a393fc4b3067f3) using the mail gem directly, in an identical way to Postal before and after this PR. It outputs two eml files, one of which has an invalid structure.

The root part of test-broken.eml is `multipart/alternative` but includes a part that represents a test PDF attachment. However, it should be a `multipart/mixed` containing a `multipart/alternative` part followed by the PDF part. The test-broken email does not render correctly on a few clients, including but not limited to:
* Outlook for Windows (no attachment visible)
* Apple Mail (attachment visible but no body rendered) (includes iOS mail app?)
* Thunderbird (I haven't personally tested this one)

It looks like this was reported to the gem [back in 2013](https://github.com/mikel/mail/issues/118) but closed by the gem author as [not a bug](https://github.com/mikel/mail/issues/118#issuecomment-12741548). The gem author wants the user to create the mixed wrapper themselves. The change by @willpower232 implements this.

We work on a similar closed-source project that parses, transforms and then outputs mime messages through a custom SMTP server. We've had similar problems with attachments and inline images in some clients (looking at you Apple Mail). Through testing, we have found the most compatible way to format the message is to always have:
```
mixed
    alternative
        text/plain
        multipart/related
            text/html
            any/images
    any/attachments
```
We always transform the input message into the above, creating parts as necessary. This ensures that all other manipulation of the message we perform doesn't result in rendering issues.

> I'd say the main thing holding this up is that we likely don't want plain text only e-mails to be within the multipart/alternative block.

If I understand correctly, @adamcooke, you would like the following behaviour after this PR is implemented:
* text part only generates a text/plain email
* html part only generates a text/html email
* both text/html generates a multipart/alternative email
* any combination of text + html that includes attachments generates a multipart/mixed with an alternative part and the attachments

I'm discounting `multipart/related` here as I'm guessing it isn't possible to embed images?

I hope this clears up what we're trying to achieve.

**Comment by mattbaylor** [comment]
This may be a naive question, however: Is this PR going to be merged? It appears to fix a bug in Postal, at least it fixes the attachment/body issue I brought up in #2236. Thanks for helping me understand the process.

**Comment by catphish** [comment]
Merged, I now understand the problem and this appears to solve it! I do apologize for taking SO long to test and merge this.

**Comment by catphish** [comment]
> Was this really merged? I tried running postal upgrade to get the latest version and I got the same version I previously had which was 2.1.2 (which has this pressing issue).

It was merged a couple of days ago. There hasn't yet been a release that contains it.

**Comment by Paz08** [comment]
Any news on the release of this fix?

**Comment by catphish** [comment]
No. The next release is awaiting the completion of https://github.com/postalserver/postal/pulls
No ETA but it's in progress.

**Pull request merged**